pubtools/_quay/utils/stepper.py

```diff
@@ -0,0 +1,354 @@
+        self._state = StepState()
+        self.results = StepResults()
+
+    @property
+    def fullname(self):
+        """Full name of class instance."""
+        return "%s:%s" % (self.NAME, self.uid)
+
+    @property
+    def state(self):
+        """Property for retreiving class state."""
+        return self._state.get()
+
+    def set_state(self, state):
+        """Set step state."""
+        self._state.set(state)
+
+    def run(self, on_update=None):
+        """Run the step code.
+
+        Step is expected to run when step state is ready, prep or error. For other
+        state running the step code is omitted. If step is in ready state,
+        _pre_run method is executed first and state is switched to prep.
+        After prep phase finishes, and skip is not set to True, _run method containing
+        all the code for running the step is executed.
+        After _run finishes, step state is set to failed, error or finished. Statistics
+        of step are update and potential results of step are stored in shared data object
+        After every change of step state, on_update callback is called if set
+        """
+        _on_update = on_update if on_update else lambda: ()
+        self._reset_stats()
+        if self.state == "ready":
+            self.stats["started"] = isodate_now()
+            self.set_state("prep")
+            self._pre_run()
+            _on_update()
+        try:
+            if self.state not in ("prep", "error"):
+                return
+            if not self.skip:
+                self.set_state("running")
+                _on_update()
+                self._run(on_update=_on_update)
+        except StepFailedError:
+            self.set_state("failed")
+        except Exception:
+            self.set_state("error")
+            raise
+        else:
+            self.set_state("finished")
+        finally:
+            self._finish_stats()
+            self._store_results()
+            _on_update()
+
+    def _pre_run(self):
+        """Execute code needed before step run.
+
+        In this method, all neccesary preparation of data can be done.
+        It can be also used to determine if step should run or not by setting
+        self.skip to True and providing self.skip_reason string with explanation.
+        """
+        pass
+
+    def _reset_stats(self):
+        self.stats = {
+            "started": None,
+            "finished": None,
+            "skip": self.skip,
+            "skip_reason": self.skip_reason,
+            "skipped": None,
+            "state": self.state,
+        }
+
+    def _finish_stats(self):
+        self.stats["finished"] = isodate_now()
+        self.stats["skipped"] = self.skip
+        self.stats["skip"] = self.skip
+        self.stats["skip_reason"] = self.skip_reason
+        self.stats["state"] = self.state
+
+    def _store_results(self):
+        self._shared_results[self.fullname] = self.results
+
+    @abc.abstractmethod
```

Comment:
Is this Python 2.6 compatible?

Reply:
I think it is https://docs.python.org/2.6/library/abc.html#module-abc